tests/js/searchtools.js

```diff
@@ -7,6 +7,23 @@ describe('Basic html theme search', function() {
       return req.responseText;
   }
 
+  function checkRanking(expectedRanking, results) {
+    let [nextExpected, ...remainingItems] = expectedRanking;
```

Comment:
Apologies if I didn't notice this earlier, but I'm a little unclear on why this function isn't expressed as something like:

```js
expect(results.map(r=>r.slice(0, 3)).toEqual(expectedRanking)
```

I think that would likely give a clearer error message in the event that the output deviated from what was expected.

Reply:
I concur.

Reply:
Good question; it was in fact intentional that I decided to begin by checking the relative positions of expected results, instead of for a precise ordering within the search results.  To explain:

If we begin introducing more of these kind of `checkRanking` test cases, and fixing/adjusting the search code over time, then precise-order assertions could become quite fragile; we'd likely need to update multiple tests during every change.

That's not a situation I'd like: I think our tests should state our general, reasonable expectations about what kind of results should rank higher than others -- without prescribing exactly where they'll appear.

Tests written that way should allow more space for implementation/search fixes, while still describing the expected behaviour of the search engine.

In the case that we reach a change/conflict where one of these tests _has_ to be updated, that's an indicator that we might need to debate the test cases we've selected and possibly drop or modify old ones -- but the intent is for that to be a rare (and discussion-worthy) scenario.

I'm not sure whether I explained that philosophy very well - let me know if that makes sense, or if I should try to find other ways to phrase it.

Reply:
Hmm, I see. I have to confess I'm not sure if I'm totally convinced: at least at this point, the test cases are simple enough that this feels a bit like overengineering. I'd personally do the simpler approach for now, then if we find we're accumulating a lot of different examples and it's becoming difficult to reason about the effect of changes at a glance, we could add something like this.

If you're determined to use this approach out of the gate, I'd add a comment explaining the reasoning.

Reply:
I think I'm reasonable determined.  It's intended to solve the today-problem of "does this fix the result ordering bug?", the middle-term problems of "describe the search ranking logic we use, and try to catch breaking changes", and the long-ish term social problem of "project `foo` wants document `bar` to rank higher; why isn't it and what are the options to address that?".

Arguably folks could use a framework like this to self-check their projects' search result quality during continuous integration and so on.  But _that_ is definitely overengineering at this point :)

So in short: yep, I'll document these ideas in the test suite (and if they're amended or removed in future as other approaches are followed, that's fine too).

Reply:
Ok! I don't feel strongly enough about this myself to continue arguing. Good enough for now, safe enough to try. 😄